[쇼핑 장바구니 1, 2단계] 페토 미션 제출합니다.

chanho0908 opened the pull request:
# 셀프 체크리스트

## 안녕하세요 제이든 !
처음 뵙겠습니다 잘 부탁드려요 !

이번 미션 절대 절대 쉽지 않았습니다 🥲

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?


## 코드 리뷰 커뮤니케이션

 <!--
 리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
 특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
 -->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->


https://github.com/user-attachments/assets/41ae719b-15ef-4b83-9cda-138e0543a1ed

chanho0908 committed:
docs:기능 요구사항 명세서 작성

donghyun81 committed:
feat: 상품 가격 객체 추가

donghyun81 committed:
feat: 상품 객체 추가

donghyun81 committed:
feat: 상품 이미지 추가

chanho0908 committed:
feat : 상품 데이터 저장소 추가

donghyun81 committed:
feat: 상품 정보 변경

donghyun81 committed:
feat: 상품 목록 화면 Ui구현

donghyun81 committed:
build: glide 의존성 추가

chanho0908 committed:
feat : manifest internet permission 추가

chanho0908 committed:
chore : 불필요한 파일 제거

chanho0908 committed:
feat : 상품 목록 데이터 소스 추상화

chanho0908 committed:
build : databinding dependency add

chanho0908 committed:
feat : Glide Image Binding Adapter 추가

chanho0908 committed:
feat : 상품 목록 리사이클러뷰 추가

chanho0908 committed:
feat : 상품 목록 리사이클러뷰 출력 기능 추가

chanho0908 committed:
feat : 리사이클러뷰 리스트 추가 기능 구현

donghyun81 committed:
feat: 상품 목록 프래그먼트 추가

chanho0908 committed:
feat : 상품 목록 프래그먼트 제거 및 액티비로 수정

chanho0908 committed:
feat : 상품 상세 화면 액션바 추가, 메인 액션바 이름 변경

chanho0908 committed:
feat : 이미지 바인딩 어댑터 Null 처리 추가

chanho0908 committed:
feat : 상품 상세 화면 추가

chanho0908 committed:
feat : 상품 더미 데이터 추가

chanho0908 committed:
feat : 리사이클러뷰 성능 향상 옵션 추가

chanho0908 committed:
feat : ktlint max line 허용 어노테이션 추가

chanho0908 committed:
feat : 상품 목록 아이템 클릭 리스너 바인딩 추가

donghyun81 committed:
feat: 장바구니 저장소 추가

donghyun81 committed:
feat: 장바구니 레이아웃 구현

donghyun81 committed:
feat: 장바구니 화면 기능 구현

donghyun81 committed:
feat: 장바구니 이동 구현

chanho0908 committed:
refactor : 장바구니 리사이클러뷰 아이템이 짤리는 문제 수정

chanho0908 committed:
feat : 장바구니 화면 뒤로가기 기능 추가

chanho0908 committed:
feat : 장바구니 추가 기능 구현

chanho0908 committed:
feat : 장바구니 제거 기능 추가

donghyun81 committed:
feat : 리사이클러뷰 스크롤 완료시 더보기 버튼 출력 기능 추가

donghyun81 committed:
feat: 더보기 기능 구현

chanho0908 committed:
refactor : 상품 더보기시 마지막 페지 계산 로직 수정

chanho0908 committed:
refactor : 장바구니 상품 제거 로직 수정

chanho0908 committed:
feat : 상품 리사이클러뷰 어댑터 상품 추가 함수 수정

chanho0908 committed:
feat : 리사이클러뷰 아이템 제거 애니메이션 추가

chanho0908 committed:
docs : 상품 목록 더보기 기능에 대한 기능 명세 추가

chanho0908 committed:
feat : 장바구니 화면 페이징 버튼 추가

chanho0908 committed:
feat : 장바구니 목록은 페이지네이션 기능 추가

chanho0908 committed:
feat : 장바구니 삭제 기능 추가

chanho0908 committed:
feat : 장바구니 아이템이 모두 삭제될시 이전 페이지로 이동 기능 추가

chanho0908 committed:
feat : 장바구니 상품이 모두 제거되었을 때 페이지 번호가 0이되는 현상 수정

chanho0908 committed:
refactor : 상품 목록 뷰홀더 네이밍 오타 수정

chanho0908 committed:
feat : 1단계 장바구니 삭제 로직 제거

chanho0908 committed:
refactor : 쇼핑 카트 더보기 버튼 가시성 데이터바인딩 적용

chanho0908 committed:
refactor : 리사이클러뷰 스크롤 리스너 마지막 위치 상수화

chanho0908 committed:
refactor : 장바구니 페이지 관리 객체 분리

chanho0908 committed:
chore : 불필요한 파일 제거

chanho0908 committed:
feat : Ui Test 확장함수 추가

chanho0908 committed:
feat : 상품 목록 및 상품 상세 화면 이동 테스트 구현

chanho0908 committed:
build : 메인 스레드 테스트 라이브러리 추가

chanho0908 committed:
test : MainViewModel Test 추가

donghyun81 committed:
test: DetailViewModelTest 추가

donghyun81 committed:
test: CartViewModelTest 추가

chanho0908 committed:
test : 페이징 객체 테스트 추가

KwonDae commented on app/src/test/java/woowacourse/shopping/view/CartViewModelTest.kt:
```diff
@@ -0,0 +1,84 @@
+        assertThat(products).hasSize(5)
+        assertThat(page).isEqualTo(1)
+        assertThat(state.nextPageEnabled).isTrue()
+        assertThat(state.previousPageEnabled).isFalse()
```
여러개를 검증할 때 junit5의 [assertAll](https://junit.org/junit5/docs/5.0.1/api/org/junit/jupiter/api/Assertions.html#assertAll-java.lang.String-org.junit.jupiter.api.function.Executable...-)을 사용해볼 수 있는데요,
기존 방법과 어떤 차이가 있는지도 찾아보시면 좋을 것 같아요.
https://stackoverflow.com/questions/40796756/assertall-vs-multiple-assertions-in-junit5

KwonDae commented on app/src/androidTest/java/woowacourse/shopping/mathcer/ProductRecyclerViewMatchers.kt:
```diff
@@ -0,0 +1,27 @@
+    fun atPositionOnView(
+        position: Int,
+        targetViewId: Int,
+    ) = object : TypeSafeMatcher<View>() {
+        override fun describeTo(description: Description) = Unit
+
+        override fun matchesSafely(view: View): Boolean {
+            val recyclerView =
+                view.rootView.findViewById<RecyclerView>(R.id.recycler_view_product)
+                    ?: return false
+            val viewHolder =
+                recyclerView.findViewHolderForAdapterPosition(position)
```
viewMatcher와 findViewHolderForAdapterPosition 를 잘 활용해주셨네요 👍

KwonDae commented on app/src/main/java/woowacourse/shopping/data/ProductStorageImpl.kt:
```diff
@@ -0,0 +1,274 @@
+
+@Suppress("ktlint:standard:max-line-length")
+class ProductStorageImpl : ProductStorage {
+    private val products = mutableMapOf<Long, Product>()
```
products 를 map으로 구현하신 배경이 있을까요?
상품번호가 중복되는 상황이 생긴다면 어떻게 될까요?

chanho0908 replied:
우선 각 상품은 고유한 아이디가 존재해야 한다고 생각했습니다.

같은 제품일지라 하더라도 판매자가 다르거나 제조 공장이 다른 경우 다른 상품으로 구분해야 할 수 있다고 생각해요 !

때문에 상품을 고유하게 식별할 수 있는 식별자가 필요하다고 생각했습니다 🧐

그래서 이 식별자는 절대 중복될 수 없다고 생각했어요 !

그렇다면 가상의 데이터 저장소에서 식별자로 특정 상품을 조회할 때
List를 사용한다면 해당하는 상품을 찾을 때 까지 리스트를 순회하므로 이에 대한 비용이 
리스트의 전체 크기만큼 소요O(n)될 수 있기 때문에

상품의 식별자를 키로 가지는 자료구조를 사용해 상품을 바로 조회할 수 있게 하기위해 사용했습니다 !

KwonDae commented on app/src/main/java/woowacourse/shopping/data/ProductStorageImpl.kt:
```diff
@@ -0,0 +1,274 @@
+import woowacourse.shopping.domain.Product
+
+@Suppress("ktlint:standard:max-line-length")
+class ProductStorageImpl : ProductStorage {
```
CartStorageImpl은 object로 만들어주셨는데,
여기는 class로 생성한 이유가 있을까요?

chanho0908 replied:
상품 목록을 저장하는 저장소는 항상 고정된 데이터를 화면에 보여주기 때문에
Class를 사용해도 괜찮았지만

`CartStorageImpl`은 사용자가 추가한 장바구니 상품을
화면이 전환되도 앱이 살아있는 동안 데이터를 유지하기 위해 싱글톤으로 만들었습니다 !

KwonDae commented on app/src/main/java/woowacourse/shopping/data/CartStorageImpl.kt:
```diff
@@ -0,0 +1,105 @@
+
+@Suppress("ktlint:standard:max-line-length")
+object CartStorageImpl : CartStorage {
+    private val cart: MutableSet<Product> =
```
현재의 요구사항은 아닐 수 있지만..?
장바구니라 생각한다면 보통 
[상품 - 수량] 으로 매핑이 될거에요.

확장성을 고려했을 때 어떻게 개선할 수 있을까요?

chanho0908 replied:
가장 먼저 든 생각은 `Map<상품,수량>` 이었어요 !

하지만 다시 고민해보니 수량 외에 매핑해야할 정보가 늘어난다면 사용할 수 없는 구조라고 생각합니다

그렇다면 확장성을 위해 별도의 객체로 만들 수 있을 것 같습니다 !

KwonDae commented on app/src/main/java/woowacourse/shopping/view/cart/vm/CartViewModelFactory.kt:
```diff
@@ -0,0 +1,17 @@
+
+class CartViewModelFactory : ViewModelProvider.Factory {
+    override fun <T : ViewModel> create(modelClass: Class<T>): T {
+        val storage: CartStorage = CartStorageImpl
```
여기서 궁금한게 있어요.

ViewModel에 필요한 repository를 주입해주기 위해 만들어주셨네요.
지금은 데이터는 내부적으로 생산해서 만들어주고 있는데요,
만약, DB로부터 불러오도록 한다면 어떻게 개선할 수 있을까요?

이전 미션에서 presenter에 dataBase 객체를 주입해주기 위해서 어떤 작업들이 있었을까요?
지금은 단순히 object를 주입해주기에 ViewModelFactory라는 클래스에서 공통으로 수행할 수 있었지만
이후에 API를 붙이거나, DB 요소를 활용하는 repository가 필요하다면 지금 구조에서는 힘들어보입니다.

chanho0908 replied:
`repository`가 무엇인지 잘 모르겠지만 😵‍💫 .. 조사해본 결과 영화관 미션으로 예를 들면
아마 데이터 소스를 추상화함으로써 Presenter의 안드로이드 의존성을 제거하는 목적으로 사용할 수 있을 것 같아요 !

이전 미션에서 저는 Presenter에 DB를 주입하기 위해 별도의 Presenter Factory class 를 만들었습니다
```
class BookingHistoryPresenterFactory() {
    fun initialize(
        context: Context,
        view: BookingHistoryContract.View,
    ): BookingHistoryContract.Presenter {
        val dao = UserDatabase.getDatabase(context).ticketDao()
        val dataSource = TicketDataSourceImpl(dao)
        val executor = DefaultMainThreadExecutor()
        return BookingHistoryPresenter(view, dataSource, executor)
    }
}
```

이 팩토리 클래스는 "Presenter에 필요한 의존성을 주입한다!" 라는 책임을 가지고 설계한 객체이기 때문에
외부의 변경으로 인해 이 클래스 내에서 변경이 일어나도 괜찮다! 라고 생각했습니다 ☺️

그리고 이번 미션에서도 `viewModel`의 생명주기에 맞춰 주입되는 인자들을 관리하기 위해선 `ViewModel Factory`를
사용해야 한다는 것을 알았습니다.

때문에 ViewModel Factory도 Presenter Factory와 동일한 개념으로 사용해볼 수 있지 않을까 싶은데요...! 🤔

아직 제가 부족해 제이든이 말씀하신 내용을 올바르게 해석하지 못한 것 같기두 하네요 🥲

혹여 틀린 내용있다면 말씀 부탁드립니다 ! ☺️

KwonDae commented on app/src/main/java/woowacourse/shopping/view/cart/adatper/CartAdapter.kt:
```diff
@@ -0,0 +1,34 @@
+        viewType: Int,
+    ): CartViewHolder {
+        val binding = ItemCartBinding.inflate(LayoutInflater.from(parent.context), parent, false)
+        return CartViewHolder(binding, handler)
```
Binding객체를 생성하는것을 CartViewHolder의 역할이라고 볼 수 있을거에요.
그렇다면 onCreateViewHolder 의 parent를 전달해서 스스로 생성하게끔 만들 수 있을까요?

다양한 ViewHolder가 생길 수 있는데, 모든 바인딩 객체를 Adapter에서 만들기보다 객체 스스로가 만들 수 있도록 할 수 있을거에요.

chanho0908 replied:
https://github.com/woowacourse/android-shopping-cart/pull/104/commits/38d34c241e925c4a99030ff2d0ee11db2c7cb6ab

[코멘트](https://github.com/woowacourse/android-shopping-cart/pull/104#discussion_r2090863534) 주신 부분과 동일하게 해당 커밋에서 수정했습니다

KwonDae commented on app/src/main/java/woowacourse/shopping/view/detail/vm/DetailViewModel.kt:
```diff
@@ -0,0 +1,26 @@
+    fun addProduct() {
+        product.value?.let {
+            cartStorage.insert(it)
+        }
```
LiveData를 사용할 때 항상 value가 null일 상황을 대비해 구현해야 할거에요.
안드로이드 개발자들이 LiveData를 사용할 때 LiveData의 nullable에 대해서 어떻게 생각해왔는지, 어떤 방법을 사용했는지
한번 찾아보시면 좋을거에요(이렇게 수정해야한다는 아닙니다)
[참고](https://stackoverflow.com/questions/46488442/is-it-possible-to-enforce-non-nullability-of-livedata-values)

KwonDae commented on app/src/main/java/woowacourse/shopping/view/main/MainActivity.kt:
```diff
@@ -0,0 +1,118 @@
+            val layoutManager = binding.recyclerViewProduct.layoutManager as? LinearLayoutManager
+
+            val visiblePosition = layoutManager?.itemCount
+            viewModel.loadProducts(visiblePosition ?: 0, PAGE_SIZE)
```
이거는 어떤 로직인가요?
클릭했을 때 아이템개수를 가져와 넘겨주는걸까요?

ViewModel도 현재 몇페이지는 알고 있을거 같은데 RecyclerView에서 직접 접근하여 가져오는 과정이 궁금해요!

chanho0908 replied:
이 부분은 제가 리사이클러뷰에 대한 이해가 부족했던 것 같습니다 😓

`Adapter class`의 오버라이드되는 `getItemCount`를 사용하면
지금처럼 불필요하게 `LayoutManager`를 사용하지 않고도 아이템의 개수를 구할 수 있습니다

또한 ViewModel도 현재 아이템의 개수를 알고 있는데 이 부분을 활용할 생각을 못했네요 🥲

KwonDae commented on app/src/main/java/woowacourse/shopping/view/main/MainActivity.kt:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    private fun observeViewModel() {
+        viewModel.products.observe(this) { value ->
```
observe에 어떤 이유로 this(Activity)를 넘길 수 있는건지 찾아보셔도 좋습니다

chanho0908 replied:
이 [자료](https://jja08111.github.io/android/live-data-observer-pattern.md/)를 통해 이해해보려 했지만모두 이해하기엔 아직 어려운 것 같습니다 😅

그래도 찾아본 내용을 요약하자면,

```
@MainThread
public void observe(
    @NonNull LifecycleOwner owner,
    @NonNull Observer<? super T> observer
) 
```

observe 함수는 첫 번째 인자로 `LifecycleOwner`를 전달받습니다

### LifecycleOwner
Android의 Lifecycle 라이브러리에서 제공하는 인터페이스로, 생명주기(Lifecycle)를 가지는 객체를 의미합니다
```
public interface LifecycleOwner {
    public val lifecycle: Lifecycle
}
```
이 인터페이스를 구현한 클래스는 getLifecycle()을 통해 Lifecycle 객체를 제공하며, 이를 통해 생명주기 상태(CREATED, STARTED, RESUMED 등)를 추적할 수 있습니다.

이 덕에 LiveData는 owner의 생명주기에 맞춰 안전하게 동작하는 것 같습니다 !

KwonDae commented on app/src/main/java/woowacourse/shopping/view/main/MainActivity.kt:
```diff
@@ -0,0 +1,118 @@
+                    binding.buttonLoad.visibility = View.GONE
+                }
+            }
+        }
```
스크롤 리스너를 통해 현재  가능하다고 판단되면 버튼을 보여주는 로직이군요..!

ViewModel 은 이미 데이터를 방출할 때 loadable을 알고 있을 것이고,
그렇다면 이전 미션에서 적용해봤던 RecyclerView에 ViewType을 이용해서 버튼을 보여줄 수 있지 않을까요?
이부분도 함께 개선해보시죠!

chanho0908 replied:
https://github.com/woowacourse/android-shopping-cart/pull/104/commits/38d34c241e925c4a99030ff2d0ee11db2c7cb6ab

해당 코멘트 적용했습니다

다시 보니 분명 커밋을 나누려 했었는데 한번에 되어있군요 ,,, 🥲

KwonDae commented on app/src/main/java/woowacourse/shopping/view/main/adapter/ProductAdapter.kt:
```diff
@@ -0,0 +1,38 @@
+        val lastPosition = items.size
+        val subList = newItems.subList(lastPosition, newItems.size)
+        items = items + subList
+        notifyItemRangeInserted(lastPosition, newItems.size)
```
단순 notifyDataSetChanged()가 아닌 Inserted 활용 좋습니다

KwonDae commented on app/src/main/java/woowacourse/shopping/view/main/MainActivity.kt:
```diff
@@ -0,0 +1,118 @@
+        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)
+        binding.lifecycleOwner = this
+        binding.adapter = productsAdapter
+        viewModel.loadProducts(0, PAGE_SIZE)
```
화면전환이 발생하면 어떤 영향이 있을까요?

chanho0908 replied:
헉...! 

viewModel은 화면이 회전되어 Activity가 파괴되어도 생명주기가 한 사이클 더 길기 때문에
데이터를 보존하게 되는데 액티비티 재생성으로 인해 화면이 다시 `onCreate`를 타면서 
`ViewModel`에 상품 목록을 요청하고 이미 존재하는 상품의 목록을 다시 불러오게 되는군요 !! 😱

chanho0908 committed:
feat : 상품 목록 더보기 버튼 가시성 데이터 바인딩 적용

chanho0908 committed:
test : FakeCartStorage 충돌 수정

KwonDae reviewed (changes requested):
1,2 단계 구현 잘해주셨습니다.
기본적으로 구조를 잘 설계해주셨는데, 생각해볼만한 부분에 코멘트를 남겨드렸으니
충분히 고민해보시고 다시 요청 부탁드립니다 :)

chanho0908 committed:
chore : 공통 기능 패키지 core 추가

chanho0908 committed:
feat : 저장소의 추가 로딩될 상품 존재 여부 판별 함수명 수정

chanho0908 committed:
feat : 추가 상품을 로딩 할 수 있는지 여부에 대한 상태 클래스 추가

chanho0908 committed:
feat : 상품 목록의 더보기 버튼을 상품 리사이클러뷰의 뷰타입으로 추가

chanho0908 committed:
refactor : 상품 상세 화면 시스템바 세팅 함수 추가

chanho0908 committed:
build : mockk gradle dependecy 추가

chanho0908 committed:
feat : Product Repository 추가

chanho0908 committed:
feat : 장바구니 Repository 추가

chanho0908 committed:
refactor : 상품 상세 화면 Repository 적용

chanho0908 committed:
feat : 상품 목록 페이징 결과 객체 추가 및 패키지 수정

chanho0908 committed:
refactor : 상품 목록 최종 페이지 판별 로직 수정

chanho0908 committed:
feat : 의존성 주입 AppContainer, App Class 추가

chanho0908 committed:
test : 메인 화면 초기 상품 목록 로딩 테스트 실패 수정

chanho0908 committed:
refactor : 다음, 이젠 페이지 클릭 리스너 수정

chanho0908 committed:
refactor : 시스템바 초기화 함수 분리

chanho0908 committed:
chore : 장바구니 adapter 패키지 오타 수정

chanho0908 committed:
chore : ViewModel에서 View의 상수에 의존하는 문제 해결

chanho0908 commented on app/src/main/java/woowacourse/shopping/view/core/base/BaseViewHolder.kt:
```diff
@@ -0,0 +1,8 @@
+import androidx.recyclerview.widget.RecyclerView
+import androidx.viewbinding.ViewBinding
+
+abstract class BaseViewHolder<out BINDING : ViewBinding>(
+    protected val binding: BINDING,
+) : RecyclerView.ViewHolder(binding.root)
```
이전 미션에서 수업 시간에  디노의 세션에서 배운 `Adapter`를 제가 이해하고 적용할 수 있는 부분까지만 사용해봤습니다
```
abstract class BaseViewHolder(
    parent: ViewGroup,
    @LayoutRes layoutRes: Int,
) : RecyclerView.ViewHolder(
        LayoutInflater.from(parent.context).inflate(layoutRes, parent, false),
    )
```

다만 이렇게 되니 Binding 객체를 직접 import하게되는데요,

이 경우 런타임에 잘못된 View를 바인딩할 가능성이 존재합니다.

```
class MovieViewHolder(
    parent: ViewGroup,
    @LayoutRes layoutId: Int,
    private val handler: Handler,
) : BaseViewHolder(parent, layoutId) {
    fun bind(item: MovieItem) {
        MovieItemBinding.bind(itemView).apply {
            model = item
            eventListener = handler
        }
    }
```

그래서 제네릭을 사용해 ViewBinding 타입을 컴파일 타임에 확정해
잘못된 Binding 클래스 사용을 원천적으로 방지 하였습니다 ☺️

chanho0908 commented on app/src/main/java/woowacourse/shopping/view/main/MainActivity.kt:
```diff
@@ -36,64 +36,70 @@ class MainActivity : AppCompatActivity(), ProductsAdapterEventHandler {
+            itemAnimator = null
+        }
+
+    private fun setSpanSizeLookup() =
+        object : GridLayoutManager.SpanSizeLookup() {
+            override fun getSpanSize(position: Int): Int {
+                return when (productsAdapter.getItemViewType(position)) {
+                    ProductRvItems.ViewType.VIEW_TYPE_PRODUCT.ordinal -> 1
+                    ProductRvItems.ViewType.VIEW_TYPE_LOAD.ordinal -> 2
+                    else -> throw IllegalArgumentException()
+                }
```
상품 목록의 더보기 버튼을 뷰홀더로 추가했습니다 !

chanho0908 reviewed (commented):
안녕하세요 제이든 !

장바구니 저장 되는 상품과 일반 상품은 서로 다른 책임과 역할을 가진 객체라고 생각해 분리하게 되었습니다.

상품 객체를 그대로 저장하는 방식에서 상품의 Id만 알도록 정보를 최소화했습니다

다만, 이런 방식에 대해 의문이 드는점이 한 가지 있습니다.
어쩌면 이런 방식이 관계형 데이터베이스의 Join을 따라하는 것 같다는 생각이 들어 
객체지향적인 관점에서 봤을 때 객체의 관계에 형성하는데 있어 적절한 구조인지 제이든의 의견을 여쭙고 싶습니다 ! ☺️

chanho0908 committed:
feat : 상품 장바구니 추가 로직 이벤트 적용

chanho0908 committed:
refactor : 상품 목록 페이지 계산시 ViewModel의 아이템 개수를 사용하도록 수정

chanho0908 committed:
refactor: LoadState를 sealed class에서 sealed interface로 변경

chanho0908 committed:
refactor : CartResult, Cart 도메인 패키지 변경

- domain/CartResult.kt -> domain/cart/CartResult.kt
- CartResult의 Product 필드 타입 Cart로 변경
- domain/cart/Cart.kt 추가

chanho0908 committed:
refactor : CartStorage의 자료구조 변경 및 CartRepository 인터페이스 수정

- CartStorage에서 Product 대신 Cart를 저장하도록 변경
- CartStorage의 cart 자료구조를 MutableSet에서 LinkedHashMap으로 변경하여 ID를 통한 접근이 가능하도록 수정
- CartRepository 인터페이스의 insert 메서드 파라미터를 Product에서 Long 타입의 productId로 변경
- CartRepository 인터페이스에 get 메서드를 추가하여 ID로 Cart를 조회할 수 있도록 기능 추가
- CartRepositoryImpl에서 위 변경사항들을 반영

chanho0908 committed:
test : 장바구니 상품 목록 테스트 수정

- 장바구니 화면에서 Product 정보를 가져오기 위한 ProductRepository 의존성 추가
- 장바구니 목록 테스트 시 상품 ID 검증 로직 추가

chanho0908 committed:
refactor : CartRepository에 Product 객체 대신 id 전달

chanho0908 committed:
test : 장바구니 테스트 픽스처 추가

chanho0908 committed:
test : 메인 화면 초기 상품 목록 로딩 테스트 수정

- ViewModel의 loadProducts() 메서드 시그니처 변경에 따라 테스트 코드 수정

chanho0908 committed:
refactor : 메인 화면 상품 목록 미리보기 추가

chanho0908 commented on app/src/main/java/woowacourse/shopping/domain/cart/Cart.kt:
```diff
@@ -0,0 +1,8 @@
+
+import java.util.UUID
+
+data class Cart(
+    val id: Long = UUID.randomUUID().mostSignificantBits,
+    val productId: Long,
+)
```
기존엔 Product 를 장바구니에서 그대로 저장하고 있습니다

이 경우 장바구니에 저장된 상품을 사용하는 곳에서 Product를 그대로 꺼내와 쓸 수 있기 때문에 편하지만,
장바구니 저장소가 상품에 대한 정보를 모두 아는것에 대한 의문이 들었습니다

장바구니 상품이 실제 상품에 대한 모든 정보가 필요하지 않을 수 있으니까요 ! ☹️
현재는 요구사항에 없지만 예를 들면 상품후기 같은 것들이요 !

그래서 실제 API를 사용하는 환경이라면 장바구니에 사용될 객체와 실제 상품 객체를 분리할 것이라고 생각했어요 ! 🙂

LocalDatabase에 장바구니에 추가한 상품의 수량과 아이디 정도만 저장하고,

실제 상품 정보가 필요한 시점에 캐싱해둔 상품 아이디로 서버에 상품 정보를 요청한다면 
괜찮은 방법이라고 생각해 탄생하게 된 아이입니다 😊

다만, 도메인에서 아이디를 직접 생성하는 것에 대해 고민해볼 내용이 많을 것 같은데요 🤔

서버-클라이언트 환경이라면 상품의 ID는 **서버**에서 내려줄 것이라고 생각했습니다
반면 장바구니 상품은 클라이언트에서 상품 목록을 전달받아 새롭게 생성하는
즉, **클라이언트에 의해 생성되는 객체**라고 생각했습니다.

그래서 도메인에서 아이디를 만들어도 합리적이라고 생각했어요 !

KwonDae replied:
> 서버-클라이언트 환경이라면 상품의 ID는 서버에서 내려줄 것이라고 생각했습니다
> 반면 장바구니 상품은 클라이언트에서 상품 목록을 전달받아 새롭게 생성하는
> 즉, 클라이언트에 의해 생성되는 객체라고 생각했습니다.

이해했습니다 !
다만, 상품 id라는 고유번호가 있는데도 클라이언트 내에서만 사용 가능한 id를 별도로 생성하는 이유를 제가 완벽하게 이해하지 못한거 같아요.
말씀하시는 의도가 외래키의 개념가 비슷한걸까요?.?

chanho0908 replied:
디엠으로 말씀드린 내용으로 대체하겠습니다 !

chanho0908 committed:
test : 다중 검증 테스트시 assertAll 적용

KwonDae commented on app/src/main/java/woowacourse/shopping/App.kt:
```diff
@@ -0,0 +1,13 @@
+    override fun onCreate() {
+        super.onCreate()
+        container = AppContainer()
+    }
```
이런 경우엔 lazy로 활용해볼 수 있지 않을까요?
AppContaine 가 onCreate 시점에서의 리소스를 필요로 하지 않다면 lazy로 활용하여 initialize에 대한 이슈를 피할수도 있구요

chanho0908 replied:
https://github.com/woowacourse/android-shopping-cart/pull/104/commits/fb0d9681f9eef216445d64e9202d1f192ed628cc

해당 코멘트 적용했습니다 !

KwonDae commented on app/src/main/java/woowacourse/shopping/AppContainer.kt:
```diff
@@ -0,0 +1,17 @@
+import woowacourse.shopping.domain.repository.ProductRepository
+import kotlin.getValue
+
+class AppContainer {
+    private val productStorage = ProductStorage()
+    private val cartStorage = CartStorage
+
+    val productRepository: ProductRepository by lazy { ProductRepositoryImpl(productStorage) }
+    val cartRepository: CartRepository by lazy { CartRepositoryImpl(cartStorage) }
+}
```
https://github.com/woowacourse/android-shopping-cart/pull/104#discussion_r2093022068

앗.. 제가 Storage를 Repository로 읽어서 착각을 했네요 ㅎㅎ..
의도를 잘 파악해주시고 잘 반영해주셨습니다. 지금은 DB나 API와 같은 요소(context를 필요로 하는)가 없어서 AppContainer가 Application에 위치 하지 않아도 문제가 되진 않지만,

이전 미션에서의 내용을 기억해보면 room을 사용할 때 context가 필요했을 거에요.
그러한 초기 세팅들을 여기서 해준다고 생각해주시면 됩니다.

KwonDae commented on app/src/main/java/woowacourse/shopping/data/storage/ProductStorage.kt:
```diff
@@ -1,10 +1,11 @@
 @Suppress("ktlint:standard:max-line-length")
-class ProductStorageImpl : ProductStorage {
+class ProductStorage {
     private val products = mutableMapOf<Long, Product>()
```
https://github.com/woowacourse/android-shopping-cart/pull/104#discussion_r2092685330

사실 이전엔 ProductStorage가 데이터생산 & 공급 두가지 역할을 모두 하고 있었지만
이제는 repository와의 연결로 "생산" 역할을 하게 되었네요 👍

KwonDae replied:
https://github.com/woowacourse/android-shopping-cart/pull/104#discussion_r2092685330

> 같은 제품일지라 하더라도 판매자가 다르거나 제조 공장이 다른 경우 다른 상품으로 구분해야 할 수 있다고 생각해요 !
> 그래서 이 식별자는 절대 중복될 수 없다고 생각했어요 !

맞습니다. 
다만, API에서 상품목록을 내려준다고 가정했을 때 클라이언트에서는 이 상품이 절대적으로 중복이 될 수 없음을 보장할 수 있을까요?
서버개발자를 100% 의존하기에는 조금의 위험성이 존재할 수 있을거 같아요.

그렇다면 우리 클라이언트 개발단에서는 이를 어떻게 방지할 수 있을까요?
또한, 지금 상태에서 중복된 Key가 들어온다면 어떤 현상이 발생할까요?

사실 현재 구현으로도 문제가 없지만 이런 고민은 현업에 가시면 많이 하게 될 부분이라 한번 쯤 고민해보시길 바라는 의도입니다 !

chanho0908 replied:
> 사실 이전엔 ProductStorage가 데이터생산 & 공급 두가지 역할을 모두 하고 있었지만
이제는 repository와의 연결로 "생산" 역할을 하게 되었네요 👍

맞습니다 ! 말씀하신 것 처럼 Storage가 두 가지 "생성"과 "공급"책임을 가지고 있다고 생각해 수정했습니다 !

>클라이언트에서는 이 상품이 절대적으로 중복이 될 수 없음을 보장할 수 있을까요?
말씀하신 상황을 고민해보니 제가 드린 의견은 이런 상황을 고려하지 않은 것 같습니다 😓

말씀하신 내용들을 남은 기간 미션을 계속 진행하면서 고민해보겠습니다 !

KwonDae commented on app/src/main/java/woowacourse/shopping/view/main/adapter/ProductAdapter.kt:
```diff
@@ -1,38 +1,70 @@
         val lastPosition = items.size
-        val subList = newItems.subList(lastPosition, newItems.size)
-        items = items + subList
+        items.clear()
```
Adapter에서 갖고 있는 Items를 초기화 하는 이유가 있을까요?
만약 기존 리스트에 대해 수정을 해야할 일이 생긴다면 가능할까요?

chanho0908 replied:
앗... LMS에 있는 강의 자료를 적용해보려다 조금 어색한 구조가 된 것 같습니다 🥲

이전 구현처럼 기존의 리스트는 유지하고 새로 들어온 데이터만 추가하는 구조로 변경했습니다 !

KwonDae commented on app/src/main/java/woowacourse/shopping/view/main/adapter/ProductAdapter.kt:
```diff
@@ -1,38 +1,70 @@
+                        handler,
+                    )
+            }
+        ) as BaseViewHolder<ViewBinding>
```
캐스팅이 필요할까요?

chanho0908 replied:
앗 `BaseViewHolder`에서 out을 사용해 공변성을 허용했으므로 불필요한 타입 캐스팅이 발생했네요 😓

```
abstract class BaseViewHolder<out BINDING : ViewBinding>()
```

chanho0908 replied:
https://github.com/woowacourse/android-shopping-cart/pull/104/commits/8bb958ff4fd3226f6648cabb9dee6bf06860d148

해당 코멘트 적용했습니다

KwonDae commented on app/src/main/java/woowacourse/shopping/view/main/MainActivity.kt:
```diff
@@ -36,64 +39,68 @@ class MainActivity : AppCompatActivity(), ProductsAdapterEventHandler {
+            productsAdapter.addProductItems(value)
+        }
+
+        viewModel.loadState.observe(this) { value ->
+            value?.let { changeLoadState(it) }
```
현재 화면전환이 발생하면 이 함수는 어떤 영향을 받을까요?
한번, 이 라인에 브레이크포인트를 걸어두고 디버깅을 해보시면
어떤 일이 발생하는지 확인하실 수 있을거에요

그리고 [이 글](https://medium.com/androiddevelopers/livedata-with-snackbar-navigation-and-other-events-the-singleliveevent-case-ac2622673150) 을 읽어보시고 어떻게 개선할 수 있는지 참고해보세요 !

chanho0908 replied:
여기서는 말씀하신 문제를 찾지 못했습니다 🥲

하지만 상품을 장바구니 버튼으로 저장 후 장바구니 화면으로 이동하는 과정에서 말씀하신 문제를 발견했습니다.

https://github.com/user-attachments/assets/80889f82-b939-45f4-bade-a981c9cd9f0f

`Activity`나 `Fragment`가 재생성되면 `observe()`가 다시 호출되고, 
LiveData는 마지막 값(이벤트)을 다시 emit해 문제가 발생했습니다 😑

때문에 라이브 데이터는 중복된 값을 방출하는 것으로 보아 Ui Event를 다루기엔 적합하지 않은 것 같습니다.

이를 해결하기 위해 Single LiveData를 사용해  이벤트를 한 번만 소비하게 만드는 구조로 변경해 해결했습니다 !

https://github.com/user-attachments/assets/fdd26301-cb57-4443-81bb-691f43206ab5

https://github.com/woowacourse/android-shopping-cart/pull/104/commits/a323fba4e6e3506e7b23255d5ca30e58d8ea5e51

KwonDae reviewed (changes requested):
빠르게 피드백 반영하시느라 고생하셨어요.
다음 단계로 넘어가기 전 짚어보고 넘어갈 부분에 코멘트를 남겨드렸습니다.
확인 후 요청 부탁 드릴게요 !

chanho0908 committed:
refactor : AppContainer 초기화 방식 변경

AppContainer를 `lateinit var`로 선언하고 `onCreate()`에서 초기화하는 방식에서 `lazy`를 사용한 초기화 방식으로 변경했습니다.

chanho0908 committed:
refactor : 어댑터 불필요한 캐스팅 제거

chanho0908 committed:
refactor : ProductUiState 추가 및 viewModel, adapter 코드 변경

MainActivity, MainViewModel, ProductAdapter 에서 상품목록과 페이지네이션 로딩 상태를 ProductUiState로 관리하도록 수정했습니다.
이에 따라, 관련 코드 및 테스트 코드를 변경했습니다.

chanho0908 committed:
feat : 이벤트 래퍼 클래스 및 SingleLiveData

chanho0908 committed:
feat : 이벤트 발생 시 한번만 소비하는 SingleLiveData 추가

- 기존 LiveData는 화면 회전 시 이벤트가 재발생하는 문제 존재
- SingleLiveData를 사용하여 이벤트가 한번만 소비되도록 수정

chanho0908 committed:
feat : 장바구니 화면 UI State 정의

chanho0908 committed:
refactor : 장바구니 화면 UiState 적용

chanho0908 commented on app/src/main/java/woowacourse/shopping/view/main/vm/MainViewModel.kt:
```diff
@@ -3,32 +3,28 @@ package woowacourse.shopping.view.main.vm
-    val products: LiveData<List<Product>> = _products
-
-    private val _loadState = MutableLiveData<LoadState>()
-    val loadState: LiveData<LoadState> = _loadState
```
리팩토링을 진행하면서 ViewModel에서 Ui에 필요한 데이터와 상태를 별도로 분리해서 관리하는 구조가
어색하다고 느껴졌습니다 🤔

만약 변경되는 상태가 많아지면 그 만큼 옵저빙해야하는 LiveData도 늘어나야 하니까요 !

그래서 Ui의 데이터와 상태를 하나의 객체로 묶어 관리해보고자 `UiState`를 만들었습니다

chanho0908 replied:
제이든 이 내용에 대해서 다시 고민해봤습니다 ...

이 코드가 과연 3,4단계 미션을 진행하면서 "이해하기 쉬운 코드""가 될 수 있을지에 대해 의문이 들었어요 🙁

제이든이 보시기에 어려운 구조라고 생각되면 과감히 UiState 제거하고 
상태를 옵저빙하는 라이브데이터를 하나씩 모두 만드는식으로 수정하겠습니다 !

예쁜 코드보다, 동료들이 읽기 쉬운 코드가 먼저니까요 !

chanho0908 committed:
refactor : CartProduct에서 Cart로 네이밍 변경 및 id 네이밍 컨벤션 통일

chanho0908 committed:
feat : CartState 추가 및 CartUiState 수정

CartUiState가 직접 Product를 가지고 있던 구조에서 CartState를 통해 Product를 참조하도록 수정했습니다.

chanho0908 committed:
refactor : 장바구니 어댑터에서 Product 대신 CartState 사용하도록 수정

chanho0908 committed:
refactor : CartState 도입하여 장바구니 상품 ID도 관리

장바구니 화면에서 상품 삭제 시, CartState를 사용하여 장바구니 ID와 상품 정보를 함께 관리하도록 수정했습니다.
이를 통해 PagingTest 및 관련 로직에서 CartState를 사용하도록 변경했습니다.

KwonDae commented on app/src/main/java/woowacourse/shopping/view/cart/vm/CartUiState.kt:
```diff
@@ -0,0 +1,6 @@
+package woowacourse.shopping.view.cart.vm
+
+data class CartUiState(
+    val carts: List<CartState> = emptyList(),
+    val pageState: PageState = PageState(),
+)
```
사실상 UiModel 역할을 하고 있군요?
ViewModel에서 두가지의 데이터를 한군데 모아서 관리하겠다는 의도는 좋습니다.
다만 기존 domain model과 다른게 어떤게 있을까? 는 아직 해소하기 어려운거 같아요.

3,4단계에서 해당 모델이 정말 필요하다면 보다 구체적인 역할을,
아니라면 배제해도 될 것 같습니다.

KwonDae reviewed (approved):
피드백 반영하시느라 고생 많으셨어요 !
미처 반영하지 못한 사항은 3,4단계에서 이어서 진행해주시면 됩니다.
남은 단계에서도 파이팅입니다 💪
